metawatch.js

```diff
@@ -3,26 +3,66 @@
 const fs = require('fs');
 const path = require('path');
 
-const watch = (targetPath, listener) => {
-  fs.readdir(targetPath, { withFileTypes: true }, (err, files) => {
-    for (const file of files) {
-      if (file.isDirectory()) {
-        const dirPath = path.join(targetPath, file.name);
-        watch(dirPath, listener);
-      }
-    }
-    fs.watch(targetPath, (event, fileName) => {
-      const filePath = path.join(targetPath, fileName);
-      try {
-        fs.stat(filePath, (err, stats) => {
-          if (stats.isDirectory()) watch(filePath, listener);
-        });
-      } catch {
-        return;
+const DEFAULT_DEBOUNCE = 10;
+
+class Watcher {
+  constructor(config = {}) {
+    this.debounce = config.debounce || DEFAULT_DEBOUNCE;
+
+    this.watchers = [];
+    this.isClosed = false;
+  }
+
+  watch(targetPath, listener) {
+    const wrappedListener = this.debounceWrapper(listener, this.debounce);
+
+    fs.readdir(targetPath, { withFileTypes: true }, (err, files) => {
+      if (this.isClosed) return;
+
+      for (const file of files) {
+        if (file.isDirectory()) {
+          const dirPath = path.join(targetPath, file.name);
+          this.watch(dirPath, listener);
+        }
       }
-      listener(event, fileName);
+      const watcher = fs.watch(targetPath, (event, fileName) => {
+        const filePath = path.join(targetPath, fileName);
+        try {
+          fs.stat(filePath, (err, stats) => {
+            if (stats.isDirectory()) {
+              this.watch(filePath, listener);
+            }
+          });
+        } catch {
+          return;
+        }
+        wrappedListener(event, fileName);
+      });
+      this.watchers.push(watcher);
     });
-  });
-};
+  }
+
+  close() {
+    this.isClosed = true;
+    this.watchers.forEach(watcher => watcher.close());
+  }
+
+  getUniqueEvent(fileName, event) {
+    return `${fileName}.${event}`;
+  }
+
+  debounceWrapper(listener, debounce) {
```

Comment:
Extract from class and rename to `debounce` and rename argument `debounce` to `debounceTimeout`, `timeout`